timeout_config.hh

```diff
@@ -10,20 +10,60 @@
 #pragma once
 
 #include "db/timeout_clock.hh"
+#include "utils/updateable_value.hh"
+#include <chrono>
+
+namespace db { class config; }
+
+class updated_timeout_config;
 
 struct timeout_config {
-    db::timeout_clock::duration read_timeout;
-    db::timeout_clock::duration write_timeout;
-    db::timeout_clock::duration range_read_timeout;
-    db::timeout_clock::duration counter_write_timeout;
-    db::timeout_clock::duration truncate_timeout;
-    db::timeout_clock::duration cas_timeout;
-    db::timeout_clock::duration other_timeout;
+    using duration_t = db::timeout_clock::duration;
+
+    static timeout_config from_updated(const updated_timeout_config&);
+
+    duration_t read_timeout;
+    duration_t write_timeout;
+    duration_t range_read_timeout;
+    duration_t counter_write_timeout;
+    duration_t truncate_timeout;
+    duration_t cas_timeout;
+    duration_t other_timeout;
 };
 
+struct updated_timeout_config {
+    using timeout_option_t = utils::updateable_value<uint32_t>;
+
+    timeout_option_t read_timeout_in_ms;
+    timeout_option_t write_timeout_in_ms;
+    timeout_option_t range_read_timeout_in_ms;
+    timeout_option_t counter_write_timeout_in_ms;
+    timeout_option_t truncate_timeout_in_ms;
+    timeout_option_t cas_timeout_in_ms;
+    timeout_option_t other_timeout_in_ms;
+
+    explicit updated_timeout_config(const db::config& cfg);
+
+    template<typename Rep, typename Period>
+    explicit updated_timeout_config(const std::chrono::duration<Rep, Period>& timeout) {
+        auto count_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(timeout).count();
+        read_timeout_in_ms = count_in_ms;
+        write_timeout_in_ms = count_in_ms;
+        range_read_timeout_in_ms = count_in_ms;
+        counter_write_timeout_in_ms = count_in_ms;
+        truncate_timeout_in_ms = count_in_ms;
+        cas_timeout_in_ms = count_in_ms;
+        other_timeout_in_ms = count_in_ms;
+    }
+
+    timeout_config::duration_t read_timeout() const;
```

Comment:
Wow. This caught my eye. As far as I know the `updateable_value<T>` has `operator()` on board, so if we do `updatable_timeout_config.read_timeout_in_ms()` it will work by returning the current value as plain `duration_t`. Thus, this helper (and the `write_timeout()` below too) are excessive, as any code that needs to get just the timeout immediate value -- it's `redis_options::get_read_timeout()` for example -- may just call `::operator()` on it and that's it.